tide_site_restriction.module

```diff
@@ -162,3 +163,75 @@ function tide_site_restriction_form_views_exposed_form_alter(&$form, FormStateIn
     }
   }
 }
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function tide_site_restriction_form_user_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  if (isset($form['field_user_site'])) {
+    $roles = Role::loadMultiple();
+    $states = [];
+    foreach ($roles as $role_id => $role) {
+      if ($role->hasPermission('bypass node access') || $role->hasPermission('administer nodes') || $role->hasPermission('bypass site restriction')) {
+        $states[] = [
+          [':input[name="roles[' . $role_id . ']"]' => ['checked' => TRUE]],
+          'or',
+          // We have drupal/role_delegation dependency, so need to check
+          // role_change name.
+          [':input[name="role_change[' . $role_id . ']"]' => ['checked' => TRUE]],
+          'or',
+        ];
+      }
+    }
+    $results = [];
+    if ($states) {
+      foreach ($states as $state) {
+        $results = array_merge($results, $state);
+      }
+      // Removes the last 'or'.
+      if (end($results) == 'or') {
+        array_pop($results);
+      }
+      $form['field_user_site']['widget']['#states'] = [
+        'invisible' => $results,
+      ];
+    }
+    $form['#validate'][] = '_tide_site_restriction_validation';
+  }
+}
+
+/**
+ * Custom validation.
+ */
+function _tide_site_restriction_validation(&$form, FormStateInterface $form_state) {
+  if (!isset($form['role_change']['#access']) || !$form['role_change']['#access']) {
+    $values = $form_state->getValue('roles');
+  }
+  else {
+    $values = array_column($form_state->getValue('role_change'), 'target_id');
+  }
+  if ($values) {
+    $roles = Role::loadMultiple($values);
+    if (empty($form_state->getValue('field_user_site'))) {
+      $result = FALSE;
+      foreach ($roles as $role) {
+        if ($role->hasPermission('bypass node access') || $role->hasPermission('administer nodes') || $role->hasPermission('bypass site restriction')) {
+          $result = TRUE;
+        }
+      }
+      if (!$result) {
+        $form_state->setErrorByName('field_user_site', t('Site Restriction field is required.'));
+      }
+    }
+  }
+}
+
+/**
+ * Implements hook_preprocess_HOOK().
+ */
+function tide_site_restriction_preprocess_fieldset(&$variables) {
```

Comment:
@vincent-gao  Considering that the Form API states property is going to be used for conditionally setting the visibility of this field it's worth moving all logic into one place. Move this functionality there and set the required property based on the role checks.

Reply:
Hey @GROwen, Thanks for your suggestion. but to be honest, I can't add `#required => TRUE or FALSE` based on role checks to wherever form_alter or form_validation. I don't know how to do that~ 😞 
one of the problems is that field_user_site `required` or `optional`, which based on role options in the form. I can't predict whether `field_user_site` was `required `or `optional`. 
I think the easy way to add an asterisk is to use hook_preprocess_HOOK based on my limited knowledge

Reply:
You would do it in the same way that you've applied the `invisible` state to the field.

Check out the docs for [drupal_process_states](https://api.drupal.org/api/drupal/core%21includes%21common.inc/function/drupal_process_states/8.7.x)

>  The following states may be applied to an element:
    enabled
    disabled
    required
    optional
    visible
    invisible
    checked
    unchecked
    expanded
    collapsed

Reply:
Thanks @GROwen,
There is an active open issue in drupal.org https://www.drupal.org/project/drupal/issues/2855139 saying #states required is not working in form API. 
yes, the same issue over here. so, I only keep `invisible` #states api.
I also tried [conditional fields](https://www.drupal.org/project/conditional_fields) module, but it seems not to fit our case.
but potentially we could extend the functionality of [conditional fields](https://www.drupal.org/project/conditional_fields) to provide what   `field based on another field functionality`.